configs/efs/config.org
#+OPTIONS: html5-fancy:t
#+HTML_DOCTYPE: html5
#+STARTUP: overview
#+TITLE: My Emacs Cofiguration
#+AUTHOR: Gil Mendes
#+EMAIL: gil00mendes@gmail.com

This is my literate configuration for [[https://www.gnu.org/software/emacs/][Emacs]].

* Pre configurations

** Tangle providence notice

#+begin_src elisp :noweb-ref tangle-providence
;; Tangled from dotfiles/configs/efs/README.org

#+end_src

*** early-init.el

#+begin_src elisp :tangle early-init.el :noweb yes
  <<tangle-providence>>
#+end_src

*** init.el

#+begin_src elisp :tangle init.el :noweb yes
  <<tangle-providence>>
#+end_src

** early-init.el

The [[file:early-init.el]] will be loaded before the "read configuration" is evaliated. Some configuration settings may have to be set at this stage but this should be used sparingly as it may be an inication of poor configuration when one has to resort too ofen to configuring at this state.

#+begin_src elist :tangle early-init.el
  (message "🥱 Loading early-init.el")

  (setq package-enable-at-startup nil)

  (setq use-package-verbose t)

  (add-hook 'before-init-hook (lambda () (message "🪝 Before init")))
  (add-hook 'after-init-hook (lambda () (message "🪝 After init")))
  (add-hook 'emacs-startup-hook (lambda () (message "🪝 Emacs startup")))
  (add-hook 'window-setup-hook (lambda () (message "🪝 Window setup")))
#+end_src

** init.el

We enable lexical binding, since some packages (e.g.: consult) will require this.

#+begin_src elisp :tangle init.el
  ;; -*- lexical-binding: t -*-

  (message "🚜 Loading init.el")
#+end_src

** Package manager

[[https://github.com/radian-software/straight.el#getting-started][Straight.el]] 🍀 is a popular package manager used to manage Emacs packages.

#+begin_src elisp :tangle init.el
  ;; https://github.com/radian-software/straight.el#getting-started
  (defvar bootstrap-version)
  (let ((bootstrap-file
	 (expand-file-name "straight/repos/straight.el/bootstrap.el" user-emacs-directory))
	(bootstrap-version 6))
    (unless (file-exists-p bootstrap-file)
      (with-current-buffer
	  (url-retrieve-synchronously
	   "https://raw.githubusercontent.com/radian-software/straight.el/develop/install.el"
	   'silent 'inhibit-cookies)
	(goto-char (point-max))
	(eval-print-last-sexp)))
    (load bootstrap-file nil 'nomessage))
#+end_src

*** Enabling the commonly-used *use-package* interface

For [[use-package-format][convenience]], we configure *straight.el* to use the *use-package* interface.

#+begin_src elisp :tangle init.el
  ;; https://github.com/radian-software/straight.el#integration-with-use-package
  (straight-use-package 'use-package)
#+end_src

** Startup

Time the startup and display the startup time after completed.

#+begin_src elisp :tangle init.el
  ;; The default is 800 kilobytes.  Measured in bytes.
  (setq gc-cons-threshold (* 50 1000 1000))

  ;; Profile emacs startup
  (add-hook 'emacs-startup-hook
	    (lambda ()
	      (message "*** Emacs loaded in %s seconds with %d garbage collection."
		       (emacs-init-time "%.2f")
		       gcs-done)))
#+end_src

Temporarily reduce garbage colletion to gain some performance boot during startup.

#+begin_src elisp :tangle init.el
  (let ((normal-gc-cons-threshold gc-cons-threshold)
      (normal-gc-cons-percentage gc-cons-percentage)
      (normal-file-name-handler-alist file-name-handler-alist)
      (init-gc-cons-threshold most-positive-fixnum)
      (init-gc-cons-percentage 0.6))
  (setq gc-cons-threshold init-gc-cons-threshold
        gc-cons-percentage init-gc-cons-percentage
        file-name-handler-alist nil)
  (add-hook 'after-init-hook
            `(lambda ()
               (setq gc-cons-threshold ,normal-gc-cons-threshold
                     gc-cons-percentage ,normal-gc-cons-percentage
                     file-name-handler-alist ',normal-file-name-handler-alist))))
#+end_src

Inhibit startup screen and messages.

#+begin_src elisp :tangle init.el
  (setq inhibit-startup-echo-area-message t)
  (setq inhibit-startup-screen t)
  (setq initial-scratch-message nil)
#+end_src

Performance tweaks. Don't load default library and use =fundamental-mode= to reduce amount of hooks.

#+begin_src elisp :tangle init.el
  (setq inhibit-default-init t)
  (setq initial-major-mode 'fundamental-mode)
#+end_src

** Server mode

Start the Emacs server from this instance so that all =emacsclient= calls are routed here.

#+begin_src elisp :tangle init.el
(server-start)
#+end_src

** ESC cancels all

#+begin_src elisp :tangle init.el
(global-set-key (kbd "<escape>") 'keyboard-escape-quit)
#+end_src

** Disable autosaving and ring bell

#+begin_src elisp :tangle init.el
  (setq auto-save-default nil
  ring-bell-function 'ignore)
#+end_src

** Winner mode

#+begin_src elisp :tangle init.el
  (winner-mode 1)
#+end_src

** Discard customizations
Emacs has a comprehensive customization system that allows configuration changes interactively.
Personally, I opt to ensure all the configuration I use for my environment is declarative.
As such, the following configuration sets the ~custom-file~ to be a random temporary file created each time Emacs starts.
This means any customizations made interactively are discarded entirely.

#+begin_src elisp :tangle init.el
(setq custom-file (make-temp-file ""))
#+end_src

** Just use 'y' or 'n' instead of 'yes' or 'no'

You'll find =yes-or-no= prompts coming up in Emacs a lot.
I'd much rather just type =y= or =n= than =yes= or =no= every time...

#+begin_src elisp :tangle init.el
(fset 'yes-or-no-p 'y-or-n-p)
#+end_src

** Confirm quit

This adds a confirmation prompt when quitting Emacs - because I'm only human.

#+begin_src elisp :tangle init.el
(setq confirm-kill-emacs 'yes-or-no-p)
#+end_src

** Set the scratch buffer string
Set the scratch buffer's initial contents to include a comment with a timestamp of creation.
Not really all that useful, but cleaner than the default comment, and I like having something there.

#+BEGIN_SRC elisp :tangle init.el
(setq initial-scratch-message (format ";; Scratch buffer - started on %s\n\n" (current-time-string)))
#+END_SRC

** UI

Disable GUI components.

#+begin_src elisp :tangle init.el
  (setq use-dialog-box nil)
  (push '(menu-bar-lines . 0) default-frame-alist)
  (push '(tool-bar-lines . 0) default-frame-alist)

  (scroll-bar-mode -1)
  (tool-bar-mode -1)
  (tooltip-mode -1)
  (set-fringe-mode 10)
  (menu-bar-mode -1)
  (blink-cursor-mode -1)
  (show-paren-mode 1)
#+end_src

* General configurations

** About me

Tell Emacs who's driving.

#+begin_src elisp :tangle init.el
  (setq user-full-name      "Gil Mendes"
        user-mail-address   "gil00mendes@gmail.com")
#+end_src

** Undo

Simple, stable linear undo with redo for Emacs.

#+begin_src elisp :tangle init.el
  ;; https://github.com/emacsmirror/undo-fu
  (use-package undo-fu
  :straight (undo-fu :type git
                     :host github
                     :repo "emacsmirror/undo-fu"))
#+end_src

** Displaying World Time

I work with people around the globe, so it's interesting to have this at hand. The command ~world-clock~, displays all configured clocks.

#+begin_src elisp :tangle init.el
(setq world-clock-list '(("America/Los_Angeles" "Pacific")
			 ("Europe/Lisbon" "Portugal")
			 ("Etc/UTC" "UTC"))
      world-clock-time-format "%a, %d %b %I:%M %p %Z")
#+end_src

** ibuffer adjustments

Adjust ibuffer to make it look more human.

#+begin_src elisp :tangle init.el
(use-package ibuffer
  :straight (:type built-in)
  :config
  (setq ibuffer-expert t
	ibuffer-display-summary nil
	ibuffer-show-empty-filter-groups nil
	ibuffer-use-header-line t
	ibuffer-formats '((mark modified read-only locked " "
				(name 30 30 :left :elide)
				" "
				(size 9 -1 :right)
				" "
				(mode 16 16 :left :elide)
				" " filename-and-process)
			  (mark " "
				(name 16 -1)
				" " filename)))
  :bind (("C-x C-b" . ibuffer)))
#+end_src

** Commenting Lines

#+begin_src elisp :tangle init.el
;; https://github.com/redguardtoo/evil-nerd-commenter
(use-package evil-nerd-commenter
  :straight (evil-nerd-commenter :type git
				 :host github
				 :repo "redguardtoo/evil-nerd-commenter")
  :bind (("C-/" . evilnc-comment-or-uncomment-lines)))
#+end_src

* Apperance

This section will deal with some of the visual trappings of Emacs. My desing goal is to arrive at a rather minimal, or rather clean design.

** Globably hightlight the current line

#+begin_src elisp :tangle init.el
  (global-hl-line-mode)
#+end_src

** Rule column indicator

Display a border at the position to indicate where the right margin is drawn.

#+begin_src elisp :tangle init.el
  ;; https://www.emacswiki.org/emacs/FillColumnIndicator
  (setq display-fill-column-indicator-column 119)
  (global-display-fill-column-indicator-mode 1)
#+end_src

** Text

*** Set the font

#+begin_src elisp :tangle init.el
  (set-face-attribute 'default nil :font "Comic Code Ligatures" :height 120)

  ;; Set the fixed pitch face
  (set-face-attribute 'fixed-pitch nil
		     :font "Comic Code Ligatures"
		     :height 120)

  ;; Set the variable pitch face
  (set-face-attribute 'variable-pitch nil
		    :font "Comic Code Ligatures")
#+end_src

*** Default text scale

For global text scaling, the [[https://github.com/purcell/default-text-scale][default-text-scale]] package can be used. Without this package, scaling may require one to resize the text in every buffer independently which is an arduous task.

#+begin_src elisp :tangle init.el
  ;; https://github.com/purcell/default-text-scale
  (use-package default-text-scale
    :straight (default-text-scale :type git
				  :host github
				  :repo "purcell/default-text-scale")
    :hook ((after-init . default-text-scale-mode)))
#+end_src

** Themes

A nice gallery of Emacs themes can be found at https://emacsthemes.com/. I based the following function on https://yannesposito.com/posts/0014-change-emacs-theme-automatically/index.html.

#+begin_src elisp :tangle init.el
  (setq doom-theme 'doom-nord-light)
  (use-package doom-themes
    :straight (doom-themes :type git
			   :host github
			   :repo "doomemacs/themes"))

  (defun g0m/auto-update-theme ()
    "depending on time use different theme"
    ;; very early => gruvbox-light, solarized-light, nord-light
    (let* ((hour (nth 2 (decode-time (current-time))))
	   (theme (cond ((<= 7 hour 8)   'doom-gruvbox-light)
			((= 9 hour)      'doom-solarized-light)
			((<= 10 hour 16) 'doom-nord-light)
			((<= 17 hour 18) 'doom-gruvbox-light)
			((<= 19 hour 22) 'doom-oceanic-next)
			(t               'doom-laserwave))))
      (when (not (equal doom-theme theme))
	(setq doom-theme theme)
	(load-theme doom-theme t))
      ;; run that function again next hour
      (run-at-time (format "%02d:%02d" (+ hour 1) 0) nil 'g0m/auto-update-theme)))

  (g0m/auto-update-theme)
#+end_src

** svg-tag-mode

#+begin_src elisp :tangle init.el
  (use-package svg-tag-mode
    :straight (svg-tag-mode :type git
			    :host github
			    :repo "rougier/svg-tag-mode"))
#+end_src

** Comment Dired-k

The standard Emacs directory editor named dired doesn't ship with convenient highlighting or coloring of the listed files. The dired-k package adds some flair to the dired ouput.

#+begin_src elisp :tangle init.el
  ;; https://github.com/emacsorphanage/dired-k
  (use-package dired-k
  :straight (dired-k :type git
                     :host github
                     :repo "emacsorphanage/dired-k")
  :init
  (setq dired-k-style 'git)
  :hook (dired-initial-position-hook . dired-k))
#+end_src

** Icons: all the icons

#+begin_src elisp :tangle init.el
  ;; https://github.com/domtronn/all-the-icons.el
  (use-package all-the-icons
    :straight (all-the-icons :type git
			     :host github
			     :repo "domtronn/all-the-icons.el"))
#+end_src

Post-installation, don't forget to run ~all-the-icons-install-fonts~.

** Scrolling

In order to [[https://www.emacswiki.org/emacs/SmoothScrolling][facilitate smoother scrolling]] than the default i.e.: "when scrolling out of view, scroll such that point is in the middle of the buffer", we set =scroll-conservatively= to allow for more line-by-line scrolling.

#+begin_src elisp :tangle init.el
  ;; https://www.emacswiki.org/emacs/SmoothScrolling
  (setq-default scroll-conservatively 100)
#+end_src

💡 If you want to center the cursor (or point in Emacs vernacular), the ~evil-scroll-line-to-center~ command bound to =z z= is your friend.

* ORG mode

Various pieces of configuration for the mighty org-mode.

#+begin_src elisp :tangle init.el
  (use-package org-mode
    :ensure nil
    :hook
    (org-mode . org-indent-mode)
    (org-mode . variable-pitch-mode)
    (org-mode . visual-line-mode)
          :custom-face
  (avy-lead             ((t (:inherit fixed-pitch))))
  (avy-lead-face-0      ((t (:inherit fixed-pitch))))
  (avy-lead-face-1      ((t (:inherit fixed-pitch))))
  (avy-lead-face-2      ((t (:inherit fixed-pitch))))
  (company-tooltip      ((t (:inherit fixed-pitch))))
  (org-table            ((t (:inherit fixed-pitch))))
  (org-formula          ((t (:inherit fixed-pitch))))
  (org-checkbox         ((t (:inherit fixed-pitch))))
  (org-code             ((t (:inherit fixed-pitch))))
  (org-verbatim         ((t (:inherit fixed-pitch))))
  (org-special-keyword  ((t (:inherit fixed-pitch))))
  (org-meta-line        ((t (:inherit fixed-pitch))))
  (org-block            ((t (:foreground nil :inherit fixed-pitch))))
  (org-indent           ((t (:inherit (org-hide fixed-pitch)))))
  (fixed-pitch          ((t (:family "Menlo" :height 120))))
  (variable-pitch       ((t (:family "ETBembo" :height 160))))

  :init
  (setq org-ellipsis " ▾"
        org-hide-emphasis-markers t
        org-src-fontify-natively t
        org-fontify-quote-and-verse-blocks t
        org-src-tab-acts-natively t
        org-edit-src-content-indentation 0
        org-src-preserve-indentation t
        org-cycle-separator-lines 2
        org-clock-persist 'history)

  (font-lock-add-keywords 'org-mode
                        '(("^ *\\([-]\\) "
                            (0 (prog1 () (compose-region (match-beginning 1) (match-end 1) "•"))))))

  :config
  (org-clock-persistence-insinuate)

  (dolist (face '((org-level-1 . 1.2)
                (org-level-2 . 1.1)
                (org-level-3 . 1.05)
                (org-level-4 . 1.0)
                (org-level-5 . 1.1)
                (org-level-6 . 1.1)
                (org-level-7 . 1.1)
                (org-level-8 . 1.1)))))
#+end_src

Add custom header indicators to make it look nicer.

#+begin_src elisp :tangle init.el
  ;; https://github.com/integral-dw/org-superstar-mode
  (use-package org-superstar
    :straight (org-superstar :type git
                             :host github
                             :repo "integral-dw/org-superstar-mode")
    :hook (org-mode . org-superstar-mode)
    :custom
    (org-superstar-remove-leading-stars t)
    (org-superstar-headline-bullets-list '("◉" "○" "●" "○" "●" "○" "●")))
#+end_src

Toggle visibility of hidden Org mode element parts upon entering and leaving an element.

#+begin_src elisp :tangle init.el
  ;; https://github.com/awth13/org-appear
  (use-package org-appear
    :straight (org-appear :type git
                          :host github
                          :repo "awth13/org-appear")
    :hook (org-mode . org-appear-mode))
#+end_src
* Development

** Projectile

[[https://github.com/bbatsov/projectile/][Projectile]] simplifies working by projects by providing some bindings that infer their behavior from a project-type. This means that we can remember single bindings expore our project trees as well as triggering [[https://docs.projectile.mx/projectile/projects.html#configure-a-projects-lifecycle-commands][project lifecycle commands]] such as configure, compile and run test, and use these generalizations across projects -- allowing ourselves to forget some project-specific details. 😌

#+begin_src elisp :tangle init.el
;; https://github.com/bbatsov/projectile/
(use-package projectile
  :straight (projectile :type git
			:host github
			:repo "bbatsov/projectile")
  :custom
  (projectile-mode-line-prefix "🗄")
  :hook (after-init .projectile-mode)
  :bind (:map projectile-mode-map
	      ("C-x p" .  projectile-command-map)))
#+end_src

** Languages

*** Parsing and syntax highlight

Enable treesit for some languages.

#+begin_src elisp :tangle init.el
(use-package treesit
  :commands (treesit-install-language-grammar g0m/treesit-install-all-languages)
  :init
  (setq treesit-language-source-alist
   '((bash . ("https://github.com/tree-sitter/tree-sitter-bash"))
     (css . ("https://github.com/tree-sitter/tree-sitter-css"))
     (dockerfile . ("https://github.com/camdencheek/tree-sitter-dockerfile"))
     (go . ("https://github.com/tree-sitter/tree-sitter-go"))
     (html . ("https://github.com/tree-sitter/tree-sitter-html"))
     (javascript . ("https://github.com/tree-sitter/tree-sitter-javascript"))
     (json . ("https://github.com/tree-sitter/tree-sitter-json"))
     (make . ("https://github.com/alemuller/tree-sitter-make"))
     (python . ("https://github.com/tree-sitter/tree-sitter-python"))
     (typescript . ("https://github.com/tree-sitter/tree-sitter-typescript" "master" "typescript/src"))
     (tsx . ("https://github.com/tree-sitter/tree-sitter-typescript" "master" "tsx/src"))
     (ruby . ("https://github.com/tree-sitter/tree-sitter-ruby"))
     (rust . ("https://github.com/tree-sitter/tree-sitter-rust"))
     (sql . ("https://github.com/m-novikov/tree-sitter-sql"))
     (toml . ("https://github.com/tree-sitter/tree-sitter-toml"))))
  :config
  (dolist (mapping '((python-mode . python-ts-mode)
		   (ruby-mode . ruby-ts-mode)
		   (css-mode . css-ts-mode)
		   (js-mode . js-ts-mode)
		   (javascript-mode . js-ts-mode)
		   (typescript-mode . typescript-ts-mode)
		   (js-json-mode . json-ts-mode)
		   (sh-mode . bash-ts-mode)))
  (add-to-list 'major-mode-remap-alist mapping))
  (add-to-list 'auto-mode-alist '("\\.[tj]sx?\\'" . tsx-ts-mode))
  (defun g0m/treesit-install-all-languages ()
    "Install all languages specified by `treesit-language-source-alist'."
    (interactive)
    (let ((languages (mapcar 'car treesit-language-source-alist)))
      (dolist (lang languages)
	      (treesit-install-language-grammar lang)
	      (message "🤟 `%s' parser was installed." lang)
	      (sit-for 0.75)))))
#+end_src

*** Language Server

A client for Language Server Protocol servers.

#+begin_src elisp :tangle init.el
(use-package eglot
  :straight (:type built-in)
  :custom
  (read-process-output-max (* 1024 1024))
  (eglot-autoshutdown t)
  :hook ((css-ts-mode-hook . eglot-ensure)
	 (html-mode-hook .eglot-ensure)
	 (js-base-mode-hook . eglot-ensure)
	 (tsx-ts-mode-hook . eglot-ensure)))
#+end_src

* Completion system

Everything related with all completion systems that are used.

** Preserve Minibuffer History with savehist-mode

#+begin_src elisp :tangle init.el
(use-package savehist
  :straight (:type built-in)
  :init
  (savehist-mode))
#+end_src

** Completions with Vertico

The [[https://github.com/minad/vertico][vertico]] package provides a lighter completion solution when compared to Helm or Ivy.

#+begin_src elisp :tangle init.el
;; https://github.com/minad/vertico
(use-package vertico
  :straight (vertico :type git
		     :host github
		     :repo "minad/vertico")
  :init
  (vertico-mode))
#+end_src

** Completions in Regions with Corfu

Corfu enhances in-buffer completion with a small completion popup.

#+begin_src elisp :tangle init.el
;; https://github.com/minad/corfu
(use-package corfu
  :straight (corfu :type git
		   :host github
		   :repo "minad/corfu")
  :bind
  (:map corfu-map
	("C-j" . corfu-next)
	("C-k" . corfu-previous)
	("TAB" . corfu-inser))
  :custom
  ;; Works with `indent-for-tab-command'. Make sure tab doesn't indent when you
  ;; want to perform completion
  (tab-always-indent 'complete)
  (corfu-auto t)
  (corfu-auto-prefix 2)
  (corfu-auto-default 0.25)
  (corfu-count 10)
  (corfu-scroll-margin 4)
  (corfu-cycle nil)
  (corfu-separator ?\s)                 ; Necessary for use with orderless
  (corfu-quit-no-match 'separator)
  (corfu-preselect-first t)             ; Preselect first candidate
  :init
  (global-corfu-mode))
#+end_src

Add support to complete at point using [[https://github.com/minad/cape][Cape]].

#+begin_src elisp :tangle init.el
;; https://github.com/minad/cape
(use-package cape
  :straight (cape :type git
		  :host github
		  :repo "minad/cape")
  :init
  (add-to-list 'completion-at-point-functions #'cape-elisp-block))
#+end_src

** Improved Candidate Filtering with Orderless

The [[https://github.com/oantolin/orderless][orderless]] package provides) more generous completion resolution by permitting us to:
1. provide partial phrases e.g.: "o i d" to filter for "org-indent-drawer" and
2. enter these parts in any order (hence /orderless/) e.g.: "drawer org" to filter for "org-indent-drawer".

#+begin_src elisp :tangle init.el
;; https://github.com/oantolin/orderless
(use-package orderless
  :straight (orderless :type git
		       :host github
		       :repo "oantolin/orderless")
  :custom
  (completion-styles '(orderless basic))

emacs_old/history
;; -*- mode: emacs-lisp; coding: utf-8-unix -*-
;; Minibuffer history file, automatically generated by ‘savehist’.

(setq savehist-minibuffer-history-variables '(buffer-name-history minibuffer-history extended-command-history file-name-history))
(setq buffer-name-history '("setup-helm.el" "init.el" "setup-appearance.el"))
(setq minibuffer-history '("/Users/gil0mendes/.emacs.d/configs/core" "/Users/gil0mendes/.emacs.d/configs/" "/Users/gil0mendes/.emacs.d/ini" "/Users/gil0mendes/.emacs.d/configs/mod" "/Users/gil0mendes/.emacs.d/configs/appe" "/Users/gil0mendes/.emacs.d/configs/" "/Users/gil0mendes/.emacs.d/configs/mode" #("setup-modeline" 0 14 (face font-lock-constant-face fontified t)) "/Users/gil0mendes/.emacs.d/ini" "/Users/gil0mendes/.emacs.d/configs/semo" "/Users/gil0mendes/.emacs.d/configs/ap" "/Users/gil0mendes/emacs" "hel" #("save-buffer" 0 11 (fontified t)) "-os" "hel" #("setup-which-key" 0 15 (face font-lock-constant-face fontified t)) #("helm-mini" 0 9 (fontified t)) "keys" "init.el" "setup-keys.el" "init" "setup-keys.el" "init" "setup-evil.el"))
(setq extended-command-history '("evil-mode" "debug" "doom-modeline-mode" "eval-buffer" "all-the-icons-install-fonts"))
(setq file-name-history '("~/.emacs.d/configs/setup-core.el" "~/.emacs.d/configs/setup-appearance.el" "~/.emacs.d/configs/setup-modeline.el" "~/.emacs.d/init.el" "~/.emacs.d/configs/setup-modeline.el" #("~/.dotfiles/emacs/configs/setup-helm.el" 1 2 (match-part "/Users/gil0mendes/.dotfiles/emacs/configs/setup-helm.el") 2 39 (match-part "/Users/gil0mendes/.dotfiles/emacs/configs/setup-helm.el")) "~/.dotfiles/emacs/configs/setup-ibuffer.el" "~/.dotfiles/config/nvim/init.vim" "~/.dotfiles/emacs/configs/setup-keys.el" "~/.dotfiles/emacs/configs/setup-which-key.el" "~/.emacs.d/configs/setup-keys.el" "~/.emacs.d/configs/setup-evil.el" "~/.emacs.d/configs/setup-projects.el" "~/.emacs.d/configs/setup-projects.el" "~/.dotfiles/emacs/configs/setup-os.el" "~/.dotfiles/emacs/configs/setup-core.el" "~/.dotfiles/emacs/configs/setup-appearance.el" "~/.dotfiles/emacs/configs/setup-os.el" "~/.dotfiles/emacs/configs/setup-core.el" "~/.dotfiles/emacs/configs/setup-core.el"))

configs/emacs-manual-config/config.el
(require 'package)
(add-to-list 'package-archives '("melpa" . "http://melpa.org/packages/"))

(unless (package-installed-p 'use-package)
  (package-refresh-contents)
  (package-install 'use-package))

(eval-when-compile
  (require 'use-package))

(use-package auto-compile
  :ensure t)

(add-hook 'emacs-startup-hook
	(lambda ()
	       (message "Loaded Emacs in %.03fs"
			(float-time (time-subtract after-init-time before-init-time)
      ))))

(let ((normal-gc-cons-threshold gc-cons-threshold)
      (normal-gc-cons-percentage gc-cons-percentage)
      (normal-file-name-handler-alist file-name-handler-alist)
      (init-gc-cons-threshold most-positive-fixnum)
      (init-gc-cons-percentage 0.6))
  (setq gc-cons-threshold init-gc-cons-threshold
        gc-cons-percentage init-gc-cons-percentage
        file-name-handler-alist nil)
  (add-hook 'after-init-hook
            `(lambda ()
               (setq gc-cons-threshold ,normal-gc-cons-threshold
                     gc-cons-percentage ,normal-gc-cons-percentage
                     file-name-handler-alist ',normal-file-name-handler-alist))))

(setq inhibit-startup-echo-area-message t)
(setq inhibit-startup-screen t)
(setq initial-scratch-message nil)

(setq inhibit-default-init t)
(setq initial-major-mode 'fundamental-mode)

(setq use-dialog-box nil)
(push '(menu-bar-lines . 0) default-frame-alist)
(push '(tool-bar-lines . 0) default-frame-alist)
(push '(vertical-scroll-bars) default-frame-alist)

(setq frame-inhibit-implied-resize t)

(advice-add #'x-apply-session-resources :override #'ignore)

(setq user-full-name    "Gil Mendes"
      user-mail-address "gil00mendes@gmail.com")

(eval-and-compile
  (defvar data-dir
    (if (getenv "XDG_DATA_HOME")
        (concat (getenv "XDG_DATA_HOME") "/emacs/")
      (expand-file-name "~/.local/share/emacs/"))
    "Directory for data.")

  (defvar cache-dir
    (if (getenv "XDG_CACHE_HOME")
        (concat (getenv "XDG_CACHE_HOME") "/emacs/")
      (expand-file-name "~/.cache/emacs/"))
    "Directory for cache.")

  (defvar pictures-dir
    (or (getenv "XDG_PICTURES_DIR")
        (expand-file-name "~/Pictures/"))
    "Directory for pictures."))

(setq use-package-always-ensure t)

(server-start)

(cond
 ((string-equal system-type "darwin")
  (setq frame-resize-pixelwise  t)
  (menu-bar-mode t)))

(setq auto-save-default       nil
      inhibit-startup-message t
      ring-bell-function      'ignore)
(dolist (mode
         '(tool-bar-mode
           tooltip-mode
           scroll-bar-mode
           blink-cursor-mode))
  (funcall mode 0))

(setq default-frame-alist
      (append (list
               '(min-height . 1)
               '(height     . 45)
               '(min-width  . 1)
               '(width      . 81)
               '(vertical-scroll-bars . nil)
               '(left-fringe    . 0)
               '(right-fringe   . 0)
               '(tool-bar-lines . 0))))

(setq-default fill-column 120)

(winner-mode 1)

(global-hl-line-mode)

(setq custom-file (make-temp-file ""))

(fset 'yes-or-no-p 'y-or-n-p)

(setq confirm-kill-emacs 'yes-or-no-p)

(setq initial-scratch-message (format ";; Scratch buffer - started on %s\n\n" (current-time-string)))

(use-package flyspell
  :ensure nil
  :bind
  (:map flyspell-mode-map ("C-;" . nil))
  :init
  (setq ispell-program-name "aspell"
        ispell-extra-args   '("--sug-mode=ultra" "--lang=en_GB")))

(use-package org-mode
  :ensure nil
  :hook
  (org-mode . org-indent-mode)
  (org-mode . variable-pitch-mode)
  (org-mode . visual-line-mode)
  :custom-face
  (avy-lead             ((t (:inherit fixed-pitch))))
  (avy-lead-face-0      ((t (:inherit fixed-pitch))))
  (avy-lead-face-1      ((t (:inherit fixed-pitch))))
  (avy-lead-face-2      ((t (:inherit fixed-pitch))))
  (company-tooltip      ((t (:inherit fixed-pitch))))
  (org-table            ((t (:inherit fixed-pitch))))
  (org-formula          ((t (:inherit fixed-pitch))))
  (org-checkbox         ((t (:inherit fixed-pitch))))
  (org-code             ((t (:inherit fixed-pitch))))
  (org-verbatim         ((t (:inherit fixed-pitch))))
  (org-special-keyword  ((t (:inherit fixed-pitch))))
  (org-meta-line        ((t (:inherit fixed-pitch))))
  (org-block            ((t (:foreground nil :inherit fixed-pitch))))
  (org-indent           ((t (:inherit (org-hide fixed-pitch)))))
  (fixed-pitch          ((t (:family "Menlo" :height 120))))
  (variable-pitch       ((t (:family "ETBembo" :height 160))))

  :init
  (setq org-ellipsis " ▾"
        org-hide-emphasis-markers t
        org-src-fontify-natively t
        org-fontify-quote-and-verse-blocks t
        org-src-tab-acts-natively t
        org-edit-src-content-indentation 0
        org-src-preserve-indentation t
        org-cycle-separator-lines 2
        org-clock-persist 'history)

  (font-lock-add-keywords 'org-mode
                          '(("^ *\\([-]\\) "
                             (0 (prog1 () (compose-region (match-beginning 1) (match-end 1) "•"))))))

  :config
  (org-clock-persistence-insinuate)

  (evil-define-key '(normal insert visual) org-mode-map (kbd "M-k") 'org-metaup)
  (evil-define-key '(normal insert visual) org-mode-map (kbd "M-j") 'org-metadown)
  (evil-define-key '(normal insert visual) org-mode-map (kbd "C-j") 'org-next-visible-heading)
  (evil-define-key '(normal insert visual) org-mode-map (kbd "C-k") 'org-previous-visible-heading)

  (dolist (face '((org-level-1 . 1.2)
                  (org-level-2 . 1.1)
                  (org-level-3 . 1.05)
                  (org-level-4 . 1.0)
                  (org-level-5 . 1.1)
                  (org-level-6 . 1.1)
                  (org-level-7 . 1.1)
                  (org-level-8 . 1.1)))))

(use-package org-superstar
  :hook (org-mode . org-superstar-mode)
  :custom
  (org-superstar-remove-leading-stars t)
  (org-superstar-headline-bullets-list '("◉" "○" "●" "○" "●" "○" "●")))

(use-package org-appear
  :hook (org-mode . org-appear-mode))

(setq comp-async-report-warnings-errors        nil
      native-comp-async-report-warnings-errors nil)

(use-package use-package)

(use-package vertico
  :hook (after-init . vertico-mode)
  :custom
  (vertico-resize t))

(use-package orderless
  :custom
  (completion-styles '(orderless))
  (completion-category-defaults nil)
  (completion-category-overrides '((file (styles partial-completion)))))

(use-package marginalia
  :hook (after-init . marginalia-mode)
  :custom
  (marginalia-annotators '(marginalia-annotators-heavy marginalia-annotators-light nil)))

(use-package embark
  :bind
  (("C-," . embark-act))

  :init
  (setq prefix-help-command #'embark-prefix-help-command))

(use-package consult
  :bind (("C-s"   . consult-line)
         ("C-c h" . consult-history)
         ("C-c m" . consult-mode-command)
         ("C-c k" . consult-kmacro)

         ("C-x M-:" . consult-complex-command)
         ("C-x b"   . consult-buffer)
         ("C-x 4 b" . consult-buffer-other-window)
         ("C-x 5 b" . consult-buffer-other-frame)
         ("C-x r b" . consult-bookmark)
         ("C-x p b" . consult-project-buffer)

         ("M-#" . consult-register-load)
         ("M-'" . consult-register-store)
         ("C-M-#" . consult-register)

         ("M-y" . consult-yank-pop)
         ("<help> a" . consult-apropos)

         ("M-g e" . consult-compile-error)
         ("M-g f" . consult-flycheck)
         ("M-g g" . consult-goto-line)
         ("M-g M-g" . consult-goto-line)
         ("M-g o" . consult-outline)
         ("M-g m" . consult-mark)
         ("M-g k" . consult-global-mark)
         ("M-g i" . consult-imenu)
         ("M-g I" . consult-imenu-multi)

         ("M-s d" . consult-find)
         ("M-s D" . consult-locate)
         ("M-s g" . consult-grep)
         ("M-s G" . consult-git-grep)
         ("M-s r" . consult-ripgrep)
         ("M-s L" . consult-line-multi)
         ("M-s m" . consult-multi-occur)
         ("M-s k" . consult-keep-lines)
         ("M-s u" . consult-focus-lines)

         ("M-s e" . consult-isearch-history)
         :map isearch-mode-map
         ("M-e" . consult-isearch-history)
         ("M-s e" . consult-isearch-history)
         ("M-s l" . consult-line)
         ("M-s L" . consult-line-multi)

         :map minibuffer-local-map
         ("M-s" . consult-history)
         ("M-r" . consult-history))

  ;; Enable automatic preview at point in the *Completions* buffer.
  :hook (completion-list-mode . consult-preview-at-point-mode)

  :init

  ;; Optionally configure the register formatting. This improves the register
  ;; preview for `consult-register', `consult-register-load',
  ;; `consult-register-store' and the Emacs built-ins.
  (setq register-preview-delay 0.5
        register-preview-function #'consult-register-format)

  ;; Add thin lines, sorting and hide the mode line of the register preview window.
  (advice-add #'register-preview :override #'consult-register-window)

  ;; Use Consult to select xref locations with preview
  (setq xref-show-xrefs-function #'consult-xref
        xref-show-definitions-function #'consult-xref)

  :config

  (consult-customize
   consult-theme
   :preview-key '(:debounce 0.2 any)
   consult-ripgrep consult-git-grep consult-grep
   consult-bookmark consult-recent-file consult-xref
   consult--source-bookmark consult--source-recent-file
   consult--source-project-recent-file
   :preview-key (kbd "M-."))

  (setq consult-narrow-key "<") ;; (kbd "C-+")

  (autoload 'projectile-project-root "projectile")
  (setq consult-project-function (lambda (_) (projectile-project-root))))
